include/cuco/static_set.cuh

```diff
@@ -201,6 +201,58 @@ class static_set {
   template <typename InputIt>
   void insert_async(InputIt first, InputIt last, cudaStream_t stream = nullptr);
 
+  /**
+   * @brief Inserts keys in the range `[first, last)` if `pred` of the corresponding stencil returns
+   * true.
+   *
+   * @note The key `*(first + i)` is inserted if `pred( *(stencil + i) )` returns true.
+   * @note This function synchronizes the given stream and returns the number of successful
+   * insertions. For asynchronous execution use `insert_if_async`.
+   *
+   * @tparam InputIt Device accessible random access iterator whose `value_type` is
+   * convertible to the container's `value_type`
+   * @tparam StencilIt Device accessible random access iterator whose value_type is
+   * convertible to Predicate's argument type
+   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
+   * argument type is convertible from <tt>std::iterator_traits<StencilIt>::value_type</tt>
+   *
+   * @param first Beginning of the sequence of key/value pairs
+   * @param last End of the sequence of key/value pairs
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
+   * @param stream CUDA stream used for the operation
+   *
+   * @return Number of successfully inserted keys
+   */
+  template <typename InputIt, typename StencilIt, typename Predicate>
+  size_type insert_if(
+    InputIt first, InputIt last, StencilIt stencil, Predicate pred, cudaStream_t stream = nullptr);
+
+  /**
+   * @brief Asynchonously inserts keys in the range `[first, last)` if `pred` of the corresponding
+   * stencil returns true.
+   *
+   * @note The key `*(first + i)` is inserted if `pred( *(stencil + i) )` returns true.
+   *
+   * @tparam InputIt Device accessible random access iterator whose `value_type` is
+   * convertible to the container's `value_type`
+   * @tparam StencilIt Device accessible random access iterator whose value_type is
+   * convertible to Predicate's argument type
+   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
+   * argument type is convertible from <tt>std::iterator_traits<StencilIt>::value_type</tt>
+   *
+   * @param first Beginning of the sequence of key/value pairs
+   * @param last End of the sequence of key/value pairs
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
+   * @param stream CUDA stream used for the operation
+   */
+  template <typename InputIt, typename StencilIt, typename Predicate>
+  void insert_if_async(
```

Comment:
If you want, you could make the async version take a reference to the value to store the number of successful insertions. That way it is the callers job to maintain the lifetime of the object.

Reply:
That would give us full flexibility, true, but I'm a bit concerned about the API complexity:
```cpp
size_type insert(InputIt, InputIt, ...); // reduction + allocation + host sync
void insert_async(InputIt, InputIt, ...); // fastest; no reduction + no allocation
void insert_async(size_type&, InputIt, InputIt, ...); // reduction + allocation
```
are three variants the user has to choose carefully from.

The only use case that I can come up with, which would benefit from the last of the aforementioned API signatures, would be the scenario where many concurrent batch `insert`s are issued to hide H2D memcopy, and we need to know if all inserts were successful. In this case, I suggest using the second API, i.e., `void insert_async(InputIt, InputIt, ...)` followed by a subsequent call to `size()` after all streams have been synced.

Reply:
For 
```cpp
void insert_async(size_type&, InputIt, InputIt, ...); // reduction + allocation
```
How would this function return the proper num. of insertions without synchronization?

Reply:
The user has to call `cudaStreamSynchronize` explicitly after the call to `insert_async` to safely use the value stored in the reference.

Reply:
Right, I take the size computation as a "bonus" of synchronization since the extra cost of doing atomic reduction is almost negligible. The fact that the async API doesn't provide size computation is intended since many use cases (e.g. in libcudf) don't need the size computation at all and we should provide a relatively less expensive code path for that purpose.